pkg/logstructured/logstructured.go

```diff
@@ -262,8 +268,85 @@ func (l *LogStructured) Update(ctx context.Context, key string, value []byte, re
 	return rev, updateEvent.KV, true, err
 }
 
+func (l *LogStructured) ttl(ctx context.Context) {
+	// vary naive TTL support
+	queue := workqueue.NewDelayingQueue()
+	ttlEventMap := &sync.Map{}
+	go func() {
+		for l.handleTTLEvents(ctx, queue, ttlEventMap) {
+		}
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			queue.ShutDown()
+			return
+		default:
+		}
+
+		for event := range l.ttlEvents(ctx) {
+			if event.Delete {
+				continue
+			}
+
+			expiredAt := time.Now().Add(time.Duration(event.KV.Lease) * time.Second)
+			eventKV := loadTTLEventKV(ttlEventMap, event.KV.Key)
+			if eventKV == nil {
+				storeTTLEventKV(ttlEventMap, event.KV.Key, event.KV.ModRevision, expiredAt)
+				queue.AddAfter(event.KV.Key, time.Duration(event.KV.Lease)*time.Second)
+			} else {
+				if event.KV.ModRevision > eventKV.modRevision {
+					storeTTLEventKV(ttlEventMap, event.KV.Key, event.KV.ModRevision, expiredAt)
+				}
+			}
```

Comment:
Question: is there any possibility of a race condition here? What happens if the queue worker removes a kv from the map in the time between it is returned from `loadTTLEventKV`, and when we check to see if it is older than the current kv? I think that should be fine?

Nits:
* simplify this by having `storeTTLEventKV` take the whole KV, instead of getting the key, revision, and ttl for it.
* make `storeTTLEventKV` return the TTL duration so that we can enqueue using it.
* enqueue the key even if there is an existing event in the queue, it's possible the new event has a lower TTL and needs to fire **before** the existing queued event (for example, 60 minute ttl updated to 5 seconds; existing queued item would fire too late).
```suggestion
			existingKV := loadTTLEventKV(ttlEventMap, event.KV.Key)
			if existingKV == nil {
				expires := storeTTLEventKV(ttlEventMap, event.KV)
				queue.AddAfter(event.KV.Key, expires)
			} else if event.KV.ModRevision > existingKV.modRevision {
				expires := storeTTLEventKV(ttlEventMap, event.KV)
				queue.AddAfter(event.KV.Key, expires)
			}
```